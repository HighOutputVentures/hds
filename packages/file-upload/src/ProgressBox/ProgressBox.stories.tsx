import { ComponentMeta, ComponentStory } from '@storybook/react';
import React from 'react';
import ProgressBox from './ProgressBox';
import { Flex, ThemeProvider } from '@highoutput/hds';

export default {
  title: 'Components/File Upload/Progress Box',
  component: ProgressBox,
} as ComponentMeta<typeof ProgressBox>;

const Template: ComponentStory<typeof ProgressBox> = (args) => (
  <ThemeProvider>
    <Flex justify="center">
      <ProgressBox {...args} />
    </Flex>
  </ThemeProvider>
);

export const Default = Template.bind({});

Default.args = {
  ...Default.args,
  value: 80,
};
