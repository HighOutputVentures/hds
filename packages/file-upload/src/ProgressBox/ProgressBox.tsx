import { Box, Flex, Icon, Progress, Text } from '@chakra-ui/react';
import {
  ErrorUploadIcon,
  SuccessCircleIcon,
  TrashIcon,
  UploadIcon,
} from '@highoutput/hds-icons';
import { FC } from 'react';

const sizes = {
  xs: 'xs',
  sm: 'sm',
  md: 'md',
  lg: 'lg',
};

export interface ProgressBox {
  size?: 'lg' | 'md' | 'sm' | 'xs';
  value?: number;
  file?: File[];
  hasError?: boolean;
  onDelete?: (data: Record<string, any>) => void;
  onTryAgain?: (data: Record<string, any>) => void;
}

const ProgressBox: FC<ProgressBox> = ({
  size = 'sm',
  value,
  file,
  hasError,
  onDelete,
  onTryAgain,
}) => {
  const calculateFileSize = (size = 0) => {
    const kilobyte = size / 1024;
    const megabyte = kilobyte / 1024;

    if (megabyte < 1) return `${kilobyte.toFixed(2)}KB`;
    return `${megabyte.toFixed(2)}MB`;
  };

  const files = (file?.length && Array.from(file)) as Record<string, any>[];
  const blob = files?.reduce((f: any) => new Blob([f], { type: f?.type }));

  const Icons = hasError ? ErrorUploadIcon : UploadIcon;
  const BGColor = hasError ? 'interface.error.500' : '';
  const BorderColor = hasError ? 'interface.error.600' : 'neutrals.200';

  return (
    <Flex
      direction="column"
      w="32rem"
      border="2px"
      bg={BGColor}
      borderColor={BorderColor}
      borderRadius="0.5rem"
      justifyContent="center"
      my={4}
    >
      <Box m={4}>
        <Flex justifyContent="space-between">
          <Flex alignItems="center">
            <Icon h="2.5rem" w="2.5rem" as={Icons} />
            <Box mx={4}>
              {hasError && (
                <Text size="label-xs-default" color={'interface.error.900'}>
                  Upload failed, please try again
                </Text>
              )}
              <Text
                fontWeight={500}
                size={!hasError ? 'paragraph-xs-default' : 'label-xs-default'}
                isTruncated
                maxW={320}
                py="1px"
                color={hasError ? 'interface.error.700' : 'neutrals.900'}
              >
                {blob?.name ?? 'Name'}
              </Text>
              {!hasError && (
                <Text size="paragraph-xs-default" color="neutrals.700">
                  {calculateFileSize(blob?.size)}
                </Text>
              )}
            </Box>
          </Flex>
          <Box
            cursor="pointer"
            onClick={onDelete}
            data-testid="hds.progress.delete.button"
          >
            <Icon
              color={!hasError ? 'neutrals.600' : 'interface.error.800'}
              h="1.25rem"
              w="1.25rem"
              as={value !== 100 ? TrashIcon : SuccessCircleIcon}
            />
          </Box>
        </Flex>
        {hasError ? (
          <Text
            mx={14}
            cursor="pointer"
            size="button-default"
            color="interface.error.800"
            onClick={onTryAgain}
          >
            Try Again
          </Text>
        ) : (
          <Flex alignItems="center">
            <Box w="100%" mr={4} my={2}>
              <Progress
                size={sizes[size]}
                isAnimated
                hasStripe
                value={value}
                data-testid="hds.progress"
                colorScheme="purple"
                borderRadius={4}
                max={100}
              />
            </Box>
            <Text
              color="neutrals.700"
              size="label-xs-default"
              data-testid="hds.progress.value"
            >
              {value}%
            </Text>
          </Flex>
        )}
      </Box>
    </Flex>
  );
};

export default ProgressBox;
