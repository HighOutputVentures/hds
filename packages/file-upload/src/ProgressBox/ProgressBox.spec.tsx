import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import React from 'react';
import ProgressBox from './ProgressBox';

describe('FileUpload', () => {
  it('should render FileUpload', () => {
    const { container } = render(<ProgressBox />);
    expect(container).toMatchSnapshot();
  });

  it('should render error progress box', () => {
    const { getByText } = render(<ProgressBox hasError />);
    expect(getByText('Upload failed, please try again')).toBeInTheDocument();
    expect(getByText('Try Again')).toBeInTheDocument();
  });

  it('should have a progress indicator', () => {
    const { getByText, getByRole } = render(<ProgressBox value={80} />);
    expect(getByRole('progressbar')).toBeInTheDocument();
    expect(getByText('80%')).toBeInTheDocument();
  });
});
