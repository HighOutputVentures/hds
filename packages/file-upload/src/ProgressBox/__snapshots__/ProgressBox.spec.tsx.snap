// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUpload should render FileUpload 1`] = `
<div>
  <div
    class="css-zg4w2o"
  >
    <div
      class="css-1kxrbkv"
    >
      <div
        class="css-gg4vpm"
      >
        <div
          class="css-70qvj9"
        >
          <svg
            class="chakra-icon chakra-icon css-1ogxkfj"
            focusable="false"
            viewBox="0 0 46 46"
          >
            <rect
              fill="#C3B8E6"
              height="40"
              rx="20"
              width="40"
              x="3"
              y="3"
            />
            <path
              d="M19.6667 26.3333L23 23M23 23L26.3333 26.3333M23 23V30.5M29.6667 26.9524C30.6846 26.1117 31.3333 24.8399 31.3333 23.4167C31.3333 20.8854 29.2813 18.8333 26.75 18.8333C26.5679 18.8333 26.3976 18.7383 26.3051 18.5814C25.2184 16.7374 23.212 15.5 20.9167 15.5C17.4649 15.5 14.6667 18.2982 14.6667 21.75C14.6667 23.4718 15.3629 25.0309 16.4891 26.1613"
              stroke="#4A3880"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.66667"
            />
            <rect
              height="40"
              rx="20"
              stroke="#EDE8FC"
              stroke-width="6"
              width="40"
              x="3"
              y="3"
            />
          </svg>
          <div
            class="css-10go69x"
          >
            <p
              class="chakra-text css-3v2y0x"
            >
              Name
            </p>
            <p
              class="chakra-text css-162yuu5"
            >
              0.00KB
            </p>
          </div>
        </div>
        <div
          class="css-4g6ai3"
        >
          <svg
            class="chakra-icon chakra-icon css-3kr7z7"
            focusable="false"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M13.3333 5.00002V4.33335C13.3333 3.39993 13.3333 2.93322 13.1517 2.5767C12.9919 2.2631 12.7369 2.00813 12.4233 1.84834C12.0668 1.66669 11.6001 1.66669 10.6667 1.66669H9.33333C8.39991 1.66669 7.9332 1.66669 7.57668 1.84834C7.26308 2.00813 7.00811 2.2631 6.84832 2.5767C6.66667 2.93322 6.66667 3.39993 6.66667 4.33335V5.00002M8.33333 9.58335V13.75M11.6667 9.58335V13.75M2.5 5.00002H17.5M15.8333 5.00002V14.3334C15.8333 15.7335 15.8333 16.4336 15.5608 16.9683C15.3212 17.4387 14.9387 17.8212 14.4683 18.0609C13.9335 18.3334 13.2335 18.3334 11.8333 18.3334H8.16667C6.76654 18.3334 6.06647 18.3334 5.53169 18.0609C5.06129 17.8212 4.67883 17.4387 4.43915 16.9683C4.16667 16.4336 4.16667 15.7335 4.16667 14.3334V5.00002"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.66667"
            />
          </svg>
        </div>
      </div>
      <div
        class="css-70qvj9"
      >
        <div
          class="css-1tachoi"
        >
          <div
            class="css-qwlwgv"
          >
            <div
              aria-valuemax="100"
              aria-valuemin="0"
              aria-valuenow="0"
              class="css-nkvdwk"
              role="progressbar"
              style="width: 0%;"
            />
          </div>
        </div>
        <p
          class="chakra-text css-162yuu5"
        >
          %
        </p>
      </div>
    </div>
  </div>
</div>
`;
